File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action7001.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System.Collections.Generic;
using ZyGames.Framework.Game.Cache;
using ZyGames.Framework.Game.Service;
using ZyGames.Framework.Common;
using ZyGames.Tianjiexing.BLL.Base;
using ZyGames.Tianjiexing.Lang;
using ZyGames.Tianjiexing.Model;
using ZyGames.Tianjiexing.Model.Enum;


namespace ZyGames.Tianjiexing.BLL.Action
{

    /// <summary>
    /// 7001_商店物品列表接口
    /// </summary>
    public class Action7001 : BaseAction
    {
        private ShopType _commandType;
        private int _pageIndex;
        private int _pageSize;
        private int _pageCount;
        private List<MallItemsInfo> _mallItemsInfoArray = new List<MallItemsInfo>();

        public Action7001(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action7001, httpGet)
        {

        }
        public override void BuildPacket()
        {
            PushIntoStack(_pageCount);
            PushIntoStack(ContextUser.GameCoin);
            PushIntoStack(ContextUser.GoldNum);
            PushIntoStack(_mallItemsInfoArray.Count);
            foreach (MallItemsInfo mallItems in _mallItemsInfoArray)
            {
                ItemBaseInfo itemInfo = new ConfigCacheSet<ItemBaseInfo>().FindKey(mallItems.ItemID);
                int mallPrice = 0;
                int specilPrice = 0;
                if (itemInfo != null)
                {
                    mallPrice = mallItems.Price;
                    if (mallItems.MallType == ShopType.Props || mallItems.MallType == ShopType.PresendBox)
                    {
                        mallPrice = (FestivalHelper.StoreDiscount() * mallPrice).ToInt();
                    }
                    specilPrice = mallItems.SpecialPrice;
                }
                DataStruct dsItem = new DataStruct();
                dsItem.PushIntoStack(mallItems.ItemID);
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.ItemName.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.HeadID.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.MaxHeadID.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.ItemDesc.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? LanguageManager.GetLang().shortInt : (short)itemInfo.QualityType);
                dsItem.PushIntoStack(mallPrice);
                dsItem.PushIntoStack(specilPrice);
                dsItem.PushIntoStack(mallItems.SeqNO);
                dsItem.PushIntoStack((short)_commandType);
                PushIntoStack(dsItem);
            }
            PushIntoStack(ContextUser.ObtainNum);
        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetEnum("MallType", ref _commandType)
                 && httpGet.GetInt("PageIndex", ref _pageIndex)
                 && httpGet.GetInt("PageSize", ref _pageSize))
            {
                return true;
            }
            return false;

        }

        public override bool TakeAction()
        {
            _mallItemsInfoArray = new ConfigCacheSet<MallItemsInfo>().FindAll(m => m.MallType == _commandType);
            _mallItemsInfoArray.GetPaging(_pageIndex, _pageSize, out _pageCount);
            return true;
        }
    }
}

File: Sample/Doudizhu/Server/release/Script/PyScript/Action/Action7001.py
import ReferenceLib
from action import *
from System import *
from mathUtils import MathUtils

from System.Collections.Generic import *
from ZyGames.Framework.SyncThreading import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Com.Rank import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Doudizhu.Bll import *
from ZyGames.Doudizhu.Bll.Logic import *
from ZyGames.Doudizhu.Bll.Com.Chat import *
from ZyGames.Framework.Game.Lang import *
from ZyGames.Doudizhu.Model import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Cache import *
from ZyGames.Doudizhu.Bll.Base import *


#商店物品列表接口
class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.ShopType = ShopType.HeadID
        self.PageIndex = 0
        self.PageSize = 0


class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)
        self.PageCount = 0
        self.GameCoin = 0
        self.GoldNum = 0
        self.ShopList = List[ShopInfo]();

def getUrlElement(httpGet, parent):
    urlParam = UrlParam()
    if httpGet.Contains("ShopType")\
    and httpGet.Contains("PageIndex")\
    and httpGet.Contains("PageSize"):
        urlParam.ShopType = httpGet.GetEnum[ShopType]("ShopType")
        urlParam.PageIndex = httpGet.GetIntValue("PageIndex")
        urlParam.PageSize = httpGet.GetIntValue("PageSize")
    else:
        urlParam.Result = False

    return urlParam

def takeAction(urlParam, parent):
    actionResult = ActionResult()
    user = parent.Current.User
    
    PaymentService.Trigger(user)
    shopList = ConfigCacheSet[ShopInfo]().FindAll(match=lambda s:s.ShopType == urlParam.ShopType)
    result = MathUtils.GetPaging[ShopInfo](shopList,urlParam.PageIndex, urlParam.PageSize)
    if result:
        actionResult.ShopList = result[0]
        actionResult.PageCount = result[1]
    actionResult.GameCoin = user.GameCoin
    gameHall = GameHall(user)
    actionResult.GoldNum = gameHall.UserGold
    #需要实现
    return actionResult


def buildPacket(writer, urlParam, actionResult):
    writer.PushIntoStack(actionResult.PageCount)
    writer.PushIntoStack(actionResult.GameCoin)
    writer.PushIntoStack(actionResult.GoldNum)
    writer.PushIntoStack(len(actionResult.ShopList))
    for info in actionResult.ShopList:
        dsItem = DataStruct()
        dsItem.PushIntoStack(info.ShopID)
        dsItem.PushIntoStack(MathUtils.ToNotNullString(info.ShopName))
        dsItem.PushIntoStack(MathUtils.ToNotNullString(info.HeadID))
        dsItem.PushIntoStack(info.Price)
        dsItem.PushIntoStack(info.VipPrice)
        dsItem.PushIntoStack(info.GameCoin)
        dsItem.PushIntoStack(MathUtils.ToNotNullString(info.ShopDesc))
        dsItem.PushIntoStack(info.SeqNO)
        writer.PushIntoStack(dsItem)
    return True